### PR TITLE
add diagnostic items for OsString/PathBuf/Owned as well as to_vec on slice

### PR DESCRIPTION
**Pull request by anall**
This is adding diagnostic items to be used by rust-lang/rust-clippy#6730, but my understanding is the clippy-side change does need to be done over there since I am adding a new clippy feature.

Add diagnostic items to the following types:
  OsString (os_string_type)
  PathBuf (path_buf_type)
  Owned (to_owned_trait)

As well as the to_vec method on slice/[T]

### CODE REVIEW

**Comment by camsteffen** [comment]
I'm hoping to resolve some questions that I [raised on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/147480-t-compiler.2Fwg-diagnostics/topic/Diagnostic.20Item.20Naming.20Convention.3F).

1. I propose that we conventionally name diagnostic item types with the exact same name as the type (`OsString` instead of `os_string_type`). This is slightly easier to read and makes for more re-usable `sym::*` entries. I don't feel strongly about this, but I would like to at least establish a convention since currently it is inconsistent.
2. I think we should generally avoid diagnostic items on associated methods and instead use diagnostic item on the type. We can find the method using `AssociatedItems::find_by_name_and_kind`. We'll have a lot less diagnostic items this way. (This applies to `slice_to_vec_method`)

File: library/alloc/src/slice.rs [hunk 1]
```diff
@@ -220,6 +220,7 @@ mod hack {
 }
 
 #[lang = "slice_alloc"]
+#[cfg_attr(not(test), rustc_diagnostic_item = "slice")]
```
Comment (anall) [inline library/alloc/src/slice.rs:223]: I'm not sure on this being `slice` -- I see the symbol was already there, but as far as I can tell wasn't being used anywhere.

Comment (camsteffen) [reply library/alloc/src/slice.rs:223]: Seems good to me. `sym::slice` is used in `rustc_hir::lang_items::LangItem::Slice::name()` but it doesn't really matter.

**Comment by anall** [comment]
And I think this is now everything I need for the clippy PR.

**Review by davidtwco — APPROVED** [review]
Apologies for the delay in getting to this review - have you been able to confirm that this is sufficient for your clippy PR?

**Pull request merged**